libsolidity/analysis/ContractLevelChecker.cpp

```diff
@@ -143,8 +178,30 @@ void ContractLevelChecker::findDuplicateDefinitions(map<string, vector<T>> const
 			SecondarySourceLocation ssl;
 
 			for (size_t j = i + 1; j < overloads.size(); ++j)
-				if (hasEqualNameAndParameters(*overloads[i], *overloads[j]))
+				if (hasEqualParameters(*overloads[i], *overloads[j]))
 				{
+					if constexpr (is_same_v<T, FunctionDefinition const*>)
+						// Once contract function aliasing is permitted, the following assertion should
+						// be removed and instead a cleaner way of asserting that either function
+						// names or function name and alias name are the same needs to be written.
+						solAssert(
+							(overloads[i]->name() == overloads[j]->name()) ||
+							(overloads[i]->isFree() && overloads[j]->isFree()),
+							""
+						);
+					else
+					{
+						static_assert(
+							is_same_v<T, EventDefinition const*>,
+							"Expected \"FunctionDefinition const*\" or \"EventDefinition const*\""
+						);
+						// Once free events are implemented, the following assertion should
```

Comment:
Do we need the distinction between events and functions at all?

Reply:
Mostly because we `isFree` is not a property of events yet, right?

Reply:
Ah, then you can maybe use `scope()` to check if it is a contract?

Reply:
Something like this 

```
				if (hasEqualParameters(*overloads[i], *overloads[j]))
				{
					solAssert(
						(
							dynamic_cast<ContractDefinition const*>(overloads[i]->scope()) &&
							dynamic_cast<ContractDefinition const*>(overloads[j]->scope())
						) ||
						(
							dynamic_cast<SourceUnit const*>(overloads[i]->scope()) &&
							dynamic_cast<SourceUnit const*>(overloads[j]->scope())
						),
						""
					);
...
```

Reply:
and what about the name?

Reply:
Something like so you mean?

```
				if (hasEqualParameters(*overloads[i], *overloads[j]))
				{
					solAssert(
						(
							dynamic_cast<ContractDefinition const*>(overloads[i]->scope()) &&
							dynamic_cast<ContractDefinition const*>(overloads[j]->scope()) &&
                                                        overloads[i]->name() == overloads[j]->name()
						) ||
						(
							dynamic_cast<SourceUnit const*>(overloads[i]->scope()) &&
							dynamic_cast<SourceUnit const*>(overloads[j]->scope())
						),
						""
					);
...
```

Reply:
The assumptions in the assertion above are the following
  - overloaded free functions/events are not guaranteed to have the same name
  - overloaded contract functions/events are guaranteed to have the same name

However, what happens when we permit aliasing of contract functions/events by means of dot limited names like `C.f`. In that case, one could do something like `import {C.f as g, g as g} from "s1.sol"` where `g` is a free function and `C.f` is a contract function both defined in file `s1.sol`. This would fail the above assertion.

Reply:
BTW, should I add a syntax test case that shows that `C.f` imports are not yet supported i.e., they lead to a parsing error?

Reply:
And yes, it is good that it fails the assertion because we probably need to put some more work into the code at that point. Yes, please add a syntax test for trying to import `C.f`.